### PR TITLE
Add optional typechecked version of constructors

### PR DESCRIPTION
**Pull request by bakkot**
(as long as you can trust typeName fields of arbitrary objects)

### CODE REVIEW

File: test/basic.js [hunk 1]
```diff
@@ -15,8 +15,42 @@
  */
 
 var ShiftAST = require('../');
+var ShiftASTChecked = require('../checked');
+var Spec = require('shift-spec').default;
 var assert = require('assert');
 
+function construct(type, checked) {
+  var AST = checked ? ShiftASTChecked : ShiftAST;
```
Comment (michaelficarra) [inline test/basic.js]: I'd rather you pass in one of these two values than a boolean.

File: test/basic.js [hunk 2]
```diff
@@ -15,8 +15,42 @@
...
+    case 'Number':
+      return 0;
+    case 'String':
+      return 'foobar';
+    case 'Maybe':
+      return null;
+    case 'List':
+      return [];
+    case 'Union':
+      return construct(type.arguments[0]);
```
Comment (michaelficarra) [inline test/basic.js]: Don't these calls to `construct` need to be passed the second parameter as the second argument?

**Comment by couchand** [comment]
As an aside, I didn't notice this PR and wrote my own version of checking the AST nodes.  It's a wrapper, rather than changing the generation directly.  Not sure if this is useful in any way here, but I thought I'd point it out.

https://github.com/couchand/shift-ast-checked

**Comment by michaelficarra** [comment]
@couchand Cool project. I hope you don't mind us obviating it with this PR. At some point, we'd like to put together a list of Shift-related projects so we don't end up duplicating effort like this.

**Comment by couchand** [comment]
It seems much cleaner to do this directly in the generated code rather than as a wrapper, so I'd be happy to see this PR merged.

I think my implementation has two features that are worth considering.  My use case is exploring the API from the REPL.  In that context it's annoying requiring the use of `new` everywhere, but that's pretty minor.

It's quite helpful to list the type that's expected rather than just throwing a generic error.  That way you don't need to switch back and forth between the REPL and the spec, and you don't have to go Ctrl-F to find the next node type in the spec.  The workflow of try-and-see is really helpful given the relatively large API surface.

File: test/basic.js (hunk 1 above)
Comment (bakkot) [inline test/basic.js]: Done.

File: test/basic.js (hunk 2 above)
Comment (bakkot) [inline test/basic.js]: Yes they do!

**Comment by bakkot** [comment]
@couchand, latest commit now reports the expected type (see e.g. [gen/checked.js#L468](https://github.com/shapesecurity/shift-ast-js/blob/f16d41becbcc49ec59d3fa2d776128682b8ca4fe/gen/checked.js#L468)). Good idea!

**Comment by bakkot** [comment]
Published as 3.1.0.

**Comment by couchand** [comment]
Nice work. Thanks! :beers:

**Pull request merged**